docs/table/indexing.rst

```diff
@@ -0,0 +1,219 @@
+.. include:: references.txt
+.. |add_index| replace:: :func:`~astropy.table.Table.add_index`
+.. |index_mode| replace:: :func:`~astropy.table.Table.index_mode`
+
+Table indexing
+--------------
+
+Once a |Table| has been created, it is possible to create indexes on one or
+more columns of the table. An index internally sorts the rows of a table based
+on the index column(s), allowing for element retrieval by column value and
+improved performance for certain table operations.
+
+Creating an index
+^^^^^^^^^^^^^^^^^
+
+To create an index on a table, use the |add_index| method::
+
+   >>> from astropy.table import Table
+   >>> t = Table([(1, 2, 3, 4), (5, 6, 7, 8)], names=('a', 'b'))
+   >>> t.add_index('a')
+
+The optional argument "unique" may be specified to create an index with
+uniquely valued elements.
+
+To create a composite index on multiple columns, pass a list of columns
+instead::
+
+   >>> t.add_index(['a', 'b'])
+
+In particular, the first index created using the
+|add_index| method is considered the default index or the "primary key". To
+retrieve an index from a table, use the `~astropy.table.Table.indices` property::
+
+   >>> t.indices['a']
+    a  rows
+   --- ----
+     1    0
+     2    1
+     3    2
+     4    3
+   >>> t.indices['a', 'b']
+    a   b  rows
+   --- --- ----
+     1   5    0
+     2   6    1
+     3   7    2
+     4   8    3
+
+
+
+Row retrieval using indices
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Row retrieval can be accomplished using two table properties: `~astropy.table.Table.loc` and
+`~astropy.table.Table.iloc`. The `~astropy.table.Table.loc` property can be indexed either by column value, range of
+column values (*including* the bounds), or a list or ndarray of column values::
+
+   >>> t = Table([(1, 2, 3, 4), (10, 1, 9, 9)], names=('a', 'b'))
+   >>> t.add_index('a')
+   >>> t.loc[2]
```

Comment:
After our discussion at the final hangout, I think this should be returning a `Row` whenever the result has length=1.